### PR TITLE
SDC: Crash fix for Questionnaire Item type equals to GROUP

### PR DESCRIPTION
**Pull request by shoaibmushtaq25**
SDC: Crash fix for Questionnaire Item type equals to GROUP
Standard data capture gallery app was crashing when Questionnaire Item type equals to GROUP.
The linked issue is https://github.com/google/android-fhir/issues/271

### CODE REVIEW

**Comment by shoaibmushtaq25** [comment]
@googlebot I signed it!

**Comment by shoaibmushtaq25** [comment]
@googlebot I fixed it.

File: datacapture/src/test/java/com/google/android/fhir/datacapture/enablement/EnablementEvaluatorTest.kt [hunk 1]
```diff
@@ -423,18 +469,39 @@ class EnablementEvaluatorTest {
...
+                    .setOperator(
+                      Questionnaire.Item.EnableWhen.OperatorCode.newBuilder()
+                        .setValue(QuestionnaireItemOperatorCode.Value.EQUALS)
+                    )
+                    .setAnswer(
+                      Questionnaire.Item.EnableWhen.AnswerX.newBuilder().apply {
+                        boolean = Boolean.newBuilder().setValue(true).build()
+                      }
+                    )
+                )
```
Comment (jingtang10) [inline datacapture/src/test/java/com/google/android/fhir/datacapture/enablement/EnablementEvaluatorTest.kt:486]: this is not needed i think? same for all the test cases below.

**Review by jingtang10 — CHANGES_REQUESTED** [review]
Thanks @shoaibmushtaq25!

File: datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt [hunk 2]
```diff
@@ -130,24 +153,24 @@ internal class QuestionnaireViewModel(state: SavedStateHandle) : ViewModel() {
...
+  private fun Questionnaire.Item.createQuestionnaireResponseItem():
+    QuestionnaireResponse.Item.Builder {
+    return QuestionnaireResponse.Item.newBuilder().apply {
+      linkId =
+        com.google.fhir.r4.core.String.newBuilder()
+          .setValue(this@createQuestionnaireResponseItem.linkId.value)
+          .build()
+      this@createQuestionnaireResponseItem.itemList.forEach {
+        this.addItem(it.createQuestionnaireResponseItem())
+      }
```
[the remark this answers is not in the document]
Comment (shoaibmushtaq25) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt]: Resolved.

**Comment by jingtang10** [comment]
please re-request review once you're done. thanks!

File: datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt [hunk 3]
```diff
@@ -41,6 +41,8 @@ internal class QuestionnaireViewModel(state: SavedStateHandle) : ViewModel() {
   /** The current questionnaire response as questions are being answered. */
   private val questionnaireResponseBuilder = QuestionnaireResponse.newBuilder()
 
+  private val questionnaireBuilder = Questionnaire.newBuilder()
```
[the remark this answers is not in the document]
Comment (shoaibmushtaq25) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt]: Resolved

File: datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt [hunk 4]
```diff
@@ -49,12 +51,18 @@ internal class QuestionnaireViewModel(state: SavedStateHandle) : ViewModel() {
     questionnaire.itemList.forEach {
       questionnaireResponseBuilder.addItem(it.createQuestionnaireResponseItem())
     }
+
+    questionnaire.itemList.forEach { questionnaireBuilder.addItem(it) }
```
[the remark this answers is not in the document]
Comment (shoaibmushtaq25) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt]: Resolved

File: datacapture/src/main/java/com/google/android/fhir/datacapture/enablement/EnablementEvaluator.kt [hunk 5]
```diff
@@ -63,7 +63,8 @@ internal object EnablementEvaluator {
    */
   fun evaluate(
     questionnaireItem: Questionnaire.Item,
-    questionnaireResponseItemRetriever: (linkId: String) -> QuestionnaireResponse.Item?
+    questionnaireResponseItemRetriever:
+      (linkId: String) -> Pair<Questionnaire.Item?, QuestionnaireResponse.Item?>
```
[the remark this answers is not in the document]
Comment (shoaibmushtaq25) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/enablement/EnablementEvaluator.kt]: Resolved, Changed to Result class.

File: datacapture/src/main/java/com/google/android/fhir/datacapture/enablement/EnablementEvaluator.kt [hunk 6]
```diff
@@ -63,7 +63,8 @@ internal object EnablementEvaluator {
    */
   fun evaluate(
     questionnaireItem: Questionnaire.Item,
-    questionnaireResponseItemRetriever: (linkId: String) -> QuestionnaireResponse.Item?
+    questionnaireResponseItemRetriever:
```
[the remark this answers is not in the document]
Comment (shoaibmushtaq25) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/enablement/EnablementEvaluator.kt]: Resolved.

File: datacapture/src/main/java/com/google/android/fhir/datacapture/enablement/EnablementEvaluator.kt [hunk 7]
```diff
@@ -85,34 +82,33 @@ internal object EnablementEvaluator {
...
+ *
+ * @param questionnaireResponseItemRetriever function that returns the [QuestionnaireResponse.Item]
+ * with the `linkId`, or null if there isn't one.
+ */
+private fun evaluateEnableWhen(
+  enableWhen: Questionnaire.Item.EnableWhen,
+  questionnaireResponseItemRetriever:
+    (linkId: String) -> Pair<Questionnaire.Item?, QuestionnaireResponse.Item?>
+): Boolean {
+  val questionnairePair = questionnaireResponseItemRetriever(enableWhen.question.value)
```
[the remark this answers is not in the document]
Comment (shoaibmushtaq25) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/enablement/EnablementEvaluator.kt]: Resolved.

File: datacapture/src/test/java/com/google/android/fhir/datacapture/enablement/EnablementEvaluatorTest.kt (hunk 1 above)
Comment (shoaibmushtaq25) [reply datacapture/src/test/java/com/google/android/fhir/datacapture/enablement/EnablementEvaluatorTest.kt:486]: I Changed the Pair to Result class here to use the power of destructive declarations. Please guide me, what should I pass here in place of QuestionnaireItem?

File: datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt [hunk 8]
```diff
@@ -62,12 +64,18 @@ internal class QuestionnaireViewModel(state: SavedStateHandle) : ViewModel() {
         questionnaireResponseBuilder.addItem(it.createQuestionnaireResponseItem())
       }
     }
+
+    //    questionnaire.itemList.forEach { questionnaireBuilder.addItem(it) }
```
Comment (jingtang10) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt:68]: ?

File: datacapture/src/test/java/com/google/android/fhir/datacapture/enablement/EnablementEvaluatorTest.kt (hunk 1 above)
Comment (jingtang10) [reply datacapture/src/test/java/com/google/android/fhir/datacapture/enablement/EnablementEvaluatorTest.kt:486]: i mean the line 474-486 the `addEnableWhen` function isn't needed. can you try without

File: datacapture/src/main/java/com/google/android/fhir/datacapture/enablement/EnablementEvaluator.kt [hunk 9]
```diff
@@ -86,33 +85,42 @@ internal object EnablementEvaluator {
...
-    questionnaireResponseItemRetriever: (linkId: String) -> QuestionnaireResponse.Item?
-  ): Boolean {
-    val responseItem = questionnaireResponseItemRetriever(enableWhen.question.value) ?: return true
-    return if (QuestionnaireItemOperatorCode.Value.EXISTS == enableWhen.operator.value) {
-      (responseItem.answerCount > 0) == enableWhen.answer.boolean.value
-    } else {
-      responseItem.contains(enableWhenTypeToPredicate(enableWhen, type))
-    }
+/** Result class to unpack questionnaireItem and questionnaireResponseItem */
+data class Result(
```
Comment (jingtang10) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/enablement/EnablementEvaluator.kt:100]:
```suggestion
data class QuestionnaireItemWithResponse(
```

or something similar?

**Review by jingtang10 — CHANGES_REQUESTED** [review]

File: datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt (hunk 8 above)
Comment (shoaibmushtaq25) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt:68]: Oh my bad, Let me remove it.

File: datacapture/src/main/java/com/google/android/fhir/datacapture/enablement/EnablementEvaluator.kt (hunk 9 above)
Comment (shoaibmushtaq25) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/enablement/EnablementEvaluator.kt]: Name changed to `QuestionnaireItemWithResponse`.

File: datacapture/src/test/java/com/google/android/fhir/datacapture/enablement/EnablementEvaluatorTest.kt (hunk 1 above)
Comment (shoaibmushtaq25) [reply datacapture/src/test/java/com/google/android/fhir/datacapture/enablement/EnablementEvaluatorTest.kt]: Yeah, removing `addEnabledWhen` worked here, so I have removed it. Thanks :+1:

**Comment by shoaibmushtaq25** [comment]
Please feel free to review this PR again @jingtang10. All comments and conflicts to date has been resolved.

**Comment by shoaibmushtaq25** [comment]
> Thanks very much!

Thanks a lot , Really appreciated.

**Pull request merged**